Fixes #38930 - Make fact parsers use proper os

ShimShtein opened the pull request:
Make sure the fact parser searches by all unique constraint combinations before creating a new object.

Steps to reproduce:
his is from memory, exact steps may be a bit different.

1. Create an OS with:

    name: RedHat
    major: 9 (or whatever matches your host OS)
    minor: 7 (or whatever matches your host OS)
    description: RHEL 9.7
2. Ensure no OS exists with name RHEL & major 9 & minor 7
3. Run the installer on a RHEL 9.7 Satellite server


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

ShimShtein committed:
Fixes #38930 - Make fact parsers use proper os

Make sure the fact parser searches by all unique constraint combinations before creating a new object.

ekohl commented on test/static_fixtures/facts/puppet_rhel_9_facter_4.1.json:
We already have [FacterDB](https://github.com/voxpupuli/facterdb) as a dependency which contains these kinds of fact sets. Is it possible to reuse that? I see others use static files so don't consider this blocking.

ShimShtein replied:
I see there is a warning under the hood of FacterDB:
```
./home/vagrant/foreman/.vendor/ruby/3.0.0/gems/jgrep-1.5.4/lib/jgrep.rb:27: warning: detected duplicate key "osfamily" in JSON object. This will raise an error in json 3.0 unless enabled via `allow_duplicate_key: true` at line 86377 column 2
/home/vagrant/foreman/.vendor/ruby/3.0.0/gems/jgrep-1.5.4/lib/jgrep.rb:27: warning: detected duplicate key "fqdn" in JSON object. This will raise an error in json 3.0 unless enabled via `allow_duplicate_key: true` at line 197772 column 2
/home/vagrant/foreman/.vendor/ruby/3.0.0/gems/jgrep-1.5.4/lib/jgrep.rb:27: warning: detected duplicate key "hostname" in JSON object. This will raise an error in json 3.0 unless enabled via `allow_duplicate_key: true` at line 402388 column 2
```
It is not problematic yet, but I preferred not to mess with external dependency.

ekohl reviewed (commented):
I didn't mean to do a review, but this caught my eye. So take my comments here as suggestions.

Looking at the failed tests I think we're using [assert_equal](https://docs.ruby-lang.org/en/2.1.0/MiniTest/Assertions.html#method-i-assert_equal) in the wrong order, leading to confusing errors. Not a blocker.

But the failing tests do look related so please have a look.

ShimShtein commented:
Fixed the original errors that @ekohl was referring to. The new errors seem trickier, since they do not appear in my local environment.

ShimShtein commented:
[test]

ShimShtein commented:
Found out that the errors came from newer gem version in the CI. Updated the gem versions locally and the errors appeared for me too. 
The errors were related to problems parsing fact files in the facterdb gem. Some of those files contain invalid json.

I have fixed the issue by forcing facterdb to load only the files that are relevant to the checks we perform. Luckily those files were OK, and the tests could run successfully.

ekohl commented:
Did you file an issue for the broken files?

ShimShtein commented:
> Did you file an issue for the broken files?

We're using an antique version of facterdb gem. In the newer versions of the gem we don't have older facter versions of the data, so I didn't test our code against the newer facterdb. If we want to file a ticket, it would not make sense to file it against v1.27 when v4.0 is already out.

stejskalleos reviewed (changes requested):
* Should the `foreman_ansible` fact parser be updated as well?
* I've never had a problem with facts & duplicate OS; Could you please share the steps on how to reproduce the issue?

ShimShtein commented:
Added ansible fact management and added the reproducer steps to the description.

Basically you want to create an operating system with the same description as the one that will be generated by the fact parser, and then run the parser.

For puppet parser it's crucial, since it is used during the upgrade procedure.

ekohl commented:
> We're using an antique version of facterdb gem. In the newer versions of the gem we don't have older facter versions of the data, so I didn't test our code against the newer facterdb. If we want to file a ticket, it would not make sense to file it against v1.27 when v4.0 is already out.

That's fair. We should look at updating, but that's for another PR.

ekohl commented on app/models/operatingsystem.rb:
```diff
@@ -46,6 +46,13 @@ class Operatingsystem < ApplicationRecord
   default_scope -> { order(:title) }
 
+  # Find all operatingsystems that are duplicates of the given operating system according to all unique constraints
+  def self.duplicate_of(os)
```
I had a hard time interpreting what `duplicate_of` meant when reading it out of context. Isn't this effectively something like `find_existing_matching_entry`?

ehelms replied:
I also find the naming not as clear, when reading it in context and would appreciate a different naming for the method.

evgeni replied:
This reads a lot like Rails' `find_or_(create|initialize)_by`, especially as in most cases you pass in a newly created object just to verify it's a dupe or not.

So couldn't we re-use that in a slightly different way (pseudocode, not actually working)?
```ruby
def self.find_or_create_by(attributes)
  self.find_by(attributes.slice(:name, :major, :minor)) || self.find_by(attributes.slice(:description)) || self.find_by(title: generate_title) || self.create(attributes)
end
```

ShimShtein replied:
Would `find_by_attributes` sound better? Or maybe something like `find_by_unique`?

As for `find_or_create` - I was thinking about it, but the code does not always creates the record immediately after the find, so I didn't want to change the behavior too much.

evgeni replied:
`find_unique_operatingsystem`? `find_operatingsystem_by_attributes`? that way the name implies this is a special thing for OSes

ehelms commented on app/models/operatingsystem.rb:
```diff
@@ -46,6 +46,13 @@ class Operatingsystem < ApplicationRecord
+  def self.duplicate_of(os)
+    scope = where(name: os.name, major: os.major, minor: os.minor)
+    scope = scope.or(where(description: os.description)) if os.description.present?
+    scope.or(where(title: os.title || os.generate_title))
```
Looks like in all examples this returns an array. I would then expect the method name to reflect this returns a plural (e.g .`duplicates_of`).

ehelms commented on app/services/foreman_salt/fact_parser.rb:
```diff
@@ -3,7 +3,8 @@ class FactParser < ::FactParser
 
     def operatingsystem
-      os = Operatingsystem.where(os_hash).first_or_initialize
+      os = Operatingsystem.new(os_hash)
```
There appears to be a pattern of calling new and then checking if a duplicate exists. I can see this being overlooked as a step when re-factoring or writing new code. Would it then make sense to include the `duplicate_of` into the `initialize` method for Operatingsystem so that it's done by default?

ShimShtein replied:
I think we should go the other way around - instead of passing a new operatingsystem object to the method, just pass a set of attributes. It makes more sense in the usage: you gather a set of attributes, and then search for an existing object that meets the defined criteria.
Maybe `find_existing` with a concrete set of attributes would be a better signature for this method.

evgeni commented on test/unit/puppet_fact_parser_test.rb:
```diff
@@ -236,6 +236,24 @@ def setup
+
+    test 'should pick up an OS by title' do
+      # Create OS with the same description as the one in the facts, but different name
+      expected = FactoryBot.create(:operatingsystem, name: 'RHEL', major: '9', minor: '6', description: 'RHEL 9.6')
+      # now we update the description in the DB to make sure we're testing the title comparison
+      Operatingsystem.where(id: expected.id).update_all(description: 'ZZZ 9.6')
```
When trying this out manually, I ended up using an OS without a description at all, should we excplicitly test that case too?

ShimShtein replied:
Since the description has a unique constraint in the DB, there would be at max one instance of such record. I think it's the same test as for any other value (that I already test).

evgeni commented on app/models/operatingsystem.rb:
```diff
@@ -46,6 +46,13 @@ class Operatingsystem < ApplicationRecord
+  def self.duplicate_of(os)
+    scope = where(name: os.name, major: os.major, minor: os.minor)
+    scope = scope.or(where(description: os.description)) if os.description.present?
+    scope.or(where(title: os.title || os.generate_title))
```
you're passing `os.title || os.generate_title` because `os` might be unsaved yet and thus `set_title` didn't run yet?

ShimShtein replied:
Yes, that's why I extracted the `generate_title` out of the `set_title` hook.

ShimShtein commented:
Updated the PR. Renamed the new search method and fixed all discrepancies associated with it.

ekohl commented on app/services/foreman_salt/fact_parser.rb:
```diff
@@ -3,7 +3,8 @@ class FactParser < ::FactParser
     def operatingsystem
-      os = Operatingsystem.where(os_hash).first_or_initialize
+      os = Operatingsystem.new(os_hash)
+      os = Operatingsystem.find_by_attributes(**os_hash).first || os
```
I'd like to avoid instantiating objects that may end up unused
```suggestion
      os = Operatingsystem.find_by_attributes(**os_hash).first || Operatingsystem.new(os_hash)
```

ekohl commented on webpack/assets/javascripts/react_app/components/Editor/__tests__/Editor.test.js:
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import { act } from '@testing-library/react';
+import { act } from 'react-dom/test-utils';
```
Is it needed to rewrite the test to another library for this bugfix?

ShimShtein replied:
It was added by mistake. Removed from the PR.

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -36,6 +36,9 @@ def operatingsystem
     end
 
+    # Make sure the operating system object that we want to create is not a duplicate of an existing one.
+    os = Operatingsystem.find_by_attributes(name: os.name, major: os.major, minor: os.minor, description: os.description).first || os
```
This is duplicating logic we already have and firing off more database queries than needed. I'd prefer to replace the calls of `find_or_initialize_by` with `find_by_attributes` and dealing with the fact it can return no instances.

ekohl commented on app/services/foreman_ansible/operating_system_parser.rb:
```diff
@@ -8,6 +8,7 @@ def operatingsystem
       args[:release_name] = os_release_name if os_name == 'Debian' || os_name == 'Ubuntu'
       # for Ansible, the CentOS Stream can be identified by missing minor version only
       args[:name] = "CentOS_Stream" if os_name == 'CentOS' && os_minor.blank?
+      args[:description] = os_description
```
This is subtly introducing a new feature. Is that intended?

ShimShtein replied:
In the original code, it was added during the generation of the new operatingsystem object. Now I have extracted it earlier, since we want to make sure no OS exists with the same description.

ShimShtein commented:
@ekohl The requirement to have only a single request in puppet fact parser caused a bit of refactoring. Hopefully I didn't change the logic too much.
Ready for another round though!

stejskalleos commented on app/models/operatingsystem.rb:
```diff
@@ -139,10 +151,26 @@ def self.families_as_collection
+  end
+
   # The OS is usually represented as the concatenation of the OS and the revision
   def to_label
```
Nit: Can we group methods like this:
```ruby
def release
end

def self.release
end
```

IMHO this can improve the readability

stejskalleos replied:
Or, we can extract these methods to the concert end then just:
```ruby
class OS
  include MethodUtils # Adds to instances
  extend MethodUtils  # Adds to the class 
end
```

stejskalleos reviewed (changes requested):
Few nits and one scenario with questionable result.

The following scenario is fixed:
* Have instance with created OS:
```
name: RedHat
major: 9
minor: 7
description: RHEL 9.7
```
* Change the name to `RHEL` (don't change description)
* Run foreman-installer

✔️ exit status 0

---

Now the broken one.

* On different host, configure foreman ansible callback
```
# ~/.ansible.cfg

[defaults]
callbacks_enabled=foreman

[callback_foreman]
url = https://foreman.example.com
verify_certs = False
ssl_cert=/etc/foreman/client_cert.pem
ssl_key = /etc/foreman/client_key.pem
client_key = /etc/foreman/client_key.pem
```
* Run ansible-playbook
* Check Foreman operating systems

🔴 There are two operating systems
1. `RHEL 9.7`, hosts: 0 (original)
2. `RedHat 9.7`, hosts: 1 (newly created)

The issue is that there should be only one OS, and the host has been reassigned to the new OS.

I'm okay with keeping it out of this scope, but we should track the issue so we won't forget it.

ShimShtein commented:
@stejskalleos 
I think the Ansible test was not indicative:
1. Ansible does not set the `description` field of the OS
2. When you change the name of the original OS, the title will also change:
```
irb(main):011:0> oo
=> 
#<Redhat:0x00007f35bd143c38
 id: 6,
 major: "9",
 name: "RedHat",
 minor: "7",
 nameindicator: nil,
 created_at: Wed, 07 Jan 2026 11:50:25.973453000 UTC +00:00,
 updated_at: Wed, 07 Jan 2026 11:50:25.973453000 UTC +00:00,
 release_name: nil,
 type: "Redhat",
 description: nil,
 password_hash: "[FILTERED]",
 title: "RedHat 9.7">
irb(main):012:0> oo.name = 'RHEL'
=> "RHEL"
irb(main):013:0> oo.save!
=> true
irb(main):014:0> oo.reload
=> 
#<Redhat:0x00007f35bd143c38
 id: 6,
 major: "9",
 name: "RHEL",
 minor: "7",
 nameindicator: nil,
 created_at: Wed, 07 Jan 2026 11:50:25.973453000 UTC +00:00,
 updated_at: Sun, 11 Jan 2026 11:16:07.835820000 UTC +00:00,
 release_name: nil,
 type: "Redhat",
 description: nil,
 password_hash: "[FILTERED]",
 title: "RHEL 9.7">
irb(main):015:0> 
```
On the other hand, I am pretty sure the OS object will be created with `Red Hat` as the name, since the only source fact that have `RHEL` in them are not related to the OS name selection:
```
irb(main):015:0> FactName.joins(:fact_values).where("lower(value) like '%rhel%'").pluck(:name, :value)
=> 
[["ansible_board_name", "RHEL"],
 ["ansible_board_version", "RHEL-9.0.0 PC (Q35 + ICH9, 2009)"],
 ["ansible_chassis_version", "RHEL-9.0.0 PC (Q35 + ICH9, 2009)"],
 ["dmi::board::product", "RHEL"],
 ["facter_dmi::board::product", "RHEL"]]

# Removed a couple of hostname-related facts that also contained the string `rhel`
```

ShimShtein commented:
Extracted the class methods into a concern. Should be a bit more aesthetic now.

stejskalleos reviewed (approved):
🍏 **LGTM**

* CI is green
* Manual verification with Packit -> OK
* Comments have been addressed
* Code LG

Since there have been multiple reviewers, I'll keep the PR open, and if no further comments are raised, I will merge it by tomorrow.

stejskalleos commented:
Thanks @ShimShtein

jeremylenz commented:
This appears to reintroduce the bug fixed in https://github.com/theforeman/foreman/pull/10425.
